File: bench/native/intloop.c
// bench: 10M-iteration integer while loop — C reference (self-timed).
#include <stdio.h>
#include <time.h>

int main(void) {
    clock_t t0 = clock();
    volatile long long total = 0; /* volatile: keep the compiler from folding the loop away */
    long long i = 0;
    while (i < 10000000) {
        total = total + i;
        i = i + 1;
    }
    double ms = (double)(clock() - t0) * 1000.0 / CLOCKS_PER_SEC;
    printf("%lld\n", (long long)total);
    fprintf(stderr, "time: %.3f ms\n", ms);
    return 0;
}

File: bench/native/intloop.rs
// bench: 10M-iteration integer while loop — Rust reference (self-timed).
fn main() {
    let t0 = std::time::Instant::now();
    let mut total: i64 = 0;
    let mut i: i64 = 0;
    while i < 10_000_000 {
        // black_box each iteration: without it LLVM folds the whole loop into the closed-form
        // sum (measured 0 ms) — the C version prevents the same folding with `volatile`.
        total = std::hint::black_box(total + i);
        i += 1;
    }
    let ms = t0.elapsed().as_secs_f64() * 1000.0;
    println!("{}", total);
    eprintln!("time: {:.3} ms", ms);
}
